gc/gc-base/src/main/java/org/projectnessie/versioned/persist/gc/GCImpl.java

```diff
@@ -0,0 +1,202 @@
+                          deletedKeys.add(operation.getKey());
+                        }
+                      });
+            }
+            handleCommit(
+                contentValuesCollector,
+                reference,
+                logEntry.getCommitMeta(),
+                deletedKeys,
+                isExpired);
+          });
+    } catch (RuntimeException e) {
+      if (e.getCause() instanceof NessieNotFoundException) {
+        LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
+      } else {
+        throw e;
+      }
+    } catch (NessieNotFoundException e) {
+      LOGGER.info("Reference {} to retrieve commits no longer exists", reference);
+    }
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void handleCommit(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      CommitMeta commitMeta,
+      Set<ContentKey> deletedKeys,
+      boolean isExpired) {
+    try {
+      Reference commitReference = referenceWithHash(reference, commitMeta);
+      List<ContentKey> allKeys =
+          api.getEntries().reference(commitReference).get().getEntries().stream()
+              .map(EntriesResponse.Entry::getName)
+              .collect(Collectors.toList());
+      api.getContent()
+          .reference(commitReference)
+          .keys(allKeys)
+          .get()
+          .forEach(
+              (contentKey, content) -> {
+                boolean isContentExpired = isExpired;
+                if (deletedKeys.contains(contentKey)) {
+                  // table is dropped. So, expire all the contents for this ContentKey independent
+                  // of cutOffTimeStamp.
+                  isContentExpired = true;
+                }
+                handleValue(
+                    contentValuesCollector, commitReference, contentKey, content, isContentExpired);
+              });
+    } catch (NessieNotFoundException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  private <GC_CONTENT_VALUES extends ExpiredContentValues> void handleValue(
+      ContentValuesCollector<GC_CONTENT_VALUES> contentValuesCollector,
+      Reference reference,
+      ContentKey contentKey,
+      Content content,
+      boolean isExpired) {
+    LOGGER.info(
+        "{} value content-id {} at commit {} in {} with key {}: {}",
+        isExpired ? "Expired" : "Live",
+        content.getId(),
+        reference.getHash(),
+        reference.getName(),
+        contentKey,
+        content);
+    contentValuesCollector.gotValue(content, reference, contentKey, isExpired);
+  }
+
+  private Reference referenceWithHash(Reference reference, CommitMeta commitMeta) {
+    if (reference instanceof Branch) {
+      reference = Branch.of(reference.getName(), commitMeta.getHash());
+    } else if (reference instanceof Tag) {
+      reference = Tag.of(reference.getName(), commitMeta.getHash());
+    } else {
```

Comment:
It might be worth adding a helper method like `withHash()` to `Reference`... I believe I saw this pattern in other places of the Nessie codebase. WDYT?